src/dql_alias.erl

```diff
@@ -0,0 +1,85 @@
+-module(dql_alias).
+
+-export([expand/2]).
+
+%%--------------------------------------------------------------------
+%% @doc Expand aliases in all query parts by replacing them with 
+%% actual selectors
+%% @end
+%%--------------------------------------------------------------------
+-spec expand([dql:statement()], [term()]) ->
+                    {error, term()} |
+                    {ok, [dql:statement()]}.
+expand(Qs, Aliases) ->
+    AliasesF =
+        lists:foldl(fun({alias, Alias, Res}, AAcc) ->
+                            gb_trees:enter(Alias, Res, AAcc)
+                    end, gb_trees:empty(), Aliases),
+    dqe_lib:pdebug('parse', "Aliases resolved.", []),
+    resolve(Qs, AliasesF).
+
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc Resolves the aliases.
+%% @end
+%%--------------------------------------------------------------------
+-spec resolve([dql:statement()], gb_trees:tree()) ->
+                        {error, term()} |
+                        {ok, [dql:statement()]}.
+resolve(Qs, Aliases) ->
+    %%{QQ, _AliasesQ} =
+    R = lists:foldl(fun(_, {error, _} = E) ->
+                            E;
+                       (Q, {QAcc, AAcc}) ->
+                            case resolve_statement(Q, AAcc) of
+                                {error, _} = E ->
+                                    E;
+                                {Q1, A1} ->
+                                    {[Q1 | QAcc], A1}
+                            end
+                    end, {[], Aliases}, Qs),
+    case R of
+        {error, _} = E ->
+            E;
+        {QQ, _AliasesQ} ->
+            dqe_lib:pdebug('parse', "Preprocessor done.", []),
+            {ok, lists:reverse(QQ)}
+    end.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc Resolves alias in a statement
+%% @end
+%%--------------------------------------------------------------------
+-spec resolve_statement(dql:statement(), gb_trees:tree()) -> 
+                               {error, term()} |
+                               {term(), gb_trees:tree()}.
+resolve_statement(O = #{op  := fcall,
+                        args := Args = #{inputs := Input}},
+                  Aliases) ->
+    {Input1, A1} =
+        lists:foldl(fun (Q, {QAcc, AAcc}) ->
+                            {Qx, Ax} = resolve_statement(Q, AAcc),
```

Comment:
In the previous function, errors returned from `resolve_statement` are explicitly handled.  It looks like errors may not not propagate to the calling function `resolve` if any errors are returned here since they will just replace the function inputs.